File: electron/database/wikiTagConnectionRepository.ts
import type Database from "better-sqlite3";
import type { WikiTagConnection } from "../types";

interface WikiTagConnectionRow {
  id: string;
  source_tag_id: string;
  target_tag_id: string;
  created_at: string;
}

function rowToConnection(row: WikiTagConnectionRow): WikiTagConnection {
  return {
    id: row.id,
    sourceTagId: row.source_tag_id,
    targetTagId: row.target_tag_id,
    createdAt: row.created_at,
  };
}

export function createWikiTagConnectionRepository(db: Database.Database) {
  const stmts = {
    fetchAll: db.prepare(
      `SELECT * FROM wiki_tag_connections ORDER BY created_at ASC`,
    ),
    fetchById: db.prepare(`SELECT * FROM wiki_tag_connections WHERE id = ?`),
    insert: db.prepare(`
      INSERT INTO wiki_tag_connections (id, source_tag_id, target_tag_id, created_at)
      VALUES (@id, @source_tag_id, @target_tag_id, @created_at)
    `),
    delete: db.prepare(`DELETE FROM wiki_tag_connections WHERE id = ?`),
    deleteByTagPair: db.prepare(`
      DELETE FROM wiki_tag_connections
      WHERE (source_tag_id = @source AND target_tag_id = @target)
         OR (source_tag_id = @target AND target_tag_id = @source)
    `),
  };

  return {
    fetchAll(): WikiTagConnection[] {
      return (stmts.fetchAll.all() as WikiTagConnectionRow[]).map(
        rowToConnection,
      );
    },

    create(sourceTagId: string, targetTagId: string): WikiTagConnection {
      const now = new Date().toISOString();
      const id = `wtc-${crypto.randomUUID()}`;
      stmts.insert.run({
        id,
        source_tag_id: sourceTagId,
        target_tag_id: targetTagId,
        created_at: now,
      });
      const row = stmts.fetchById.get(id) as WikiTagConnectionRow;
      return rowToConnection(row);
    },

    delete(id: string): void {
      stmts.delete.run(id);
    },

    deleteByTagPair(sourceTagId: string, targetTagId: string): void {
      stmts.deleteByTagPair.run({
        source: sourceTagId,
        target: targetTagId,
      });
    },
  };
}

export type WikiTagConnectionRepository = ReturnType<
  typeof createWikiTagConnectionRepository
>;

File: src-tauri/src/db/wiki_tag_connection_repository.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};

use super::helpers;
use super::row_converter::{query_all, query_one, FromRow};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WikiTagConnection {
    pub id: String,
    pub source_tag_id: String,
    pub target_tag_id: String,
    pub created_at: String,
    pub updated_at: String,
}

impl FromRow for WikiTagConnection {
    fn from_row(row: &rusqlite::Row) -> rusqlite::Result<Self> {
        Ok(WikiTagConnection {
            id: row.get("id")?,
            source_tag_id: row.get("source_tag_id")?,
            target_tag_id: row.get("target_tag_id")?,
            created_at: row.get("created_at")?,
            updated_at: row.get("updated_at")?,
        })
    }
}

pub fn fetch_all(conn: &Connection) -> rusqlite::Result<Vec<WikiTagConnection>> {
    query_all(
        conn,
        "SELECT * FROM wiki_tag_connections ORDER BY created_at",
        [],
    )
}

pub fn create(
    conn: &Connection,
    source_tag_id: &str,
    target_tag_id: &str,
) -> rusqlite::Result<WikiTagConnection> {
    let id = format!("wtc-{}", helpers::new_uuid());
    let now = helpers::now();
    conn.execute(
        "INSERT INTO wiki_tag_connections (id, source_tag_id, target_tag_id, created_at, updated_at) \
         VALUES (?1, ?2, ?3, ?4, ?5)",
        params![&id, source_tag_id, target_tag_id, &now, &now],
    )?;

    query_one(conn, "SELECT * FROM wiki_tag_connections WHERE id = ?1", [&id])
}

pub fn delete(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM wiki_tag_connections WHERE id = ?1", [id])?;
    Ok(())
}

pub fn delete_by_tag_pair(
    conn: &Connection,
    source_tag_id: &str,
    target_tag_id: &str,
) -> rusqlite::Result<()> {
    conn.execute(
        "DELETE FROM wiki_tag_connections \
         WHERE (source_tag_id = ?1 AND target_tag_id = ?2) \
         OR (source_tag_id = ?2 AND target_tag_id = ?1)",
        params![source_tag_id, target_tag_id],
    )?;
    Ok(())
}
